scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroup.scala

```diff
@@ -0,0 +1,395 @@
+import java.net.{ConnectException, InetSocketAddress}
+import java.nio.channels.ClosedChannelException
+import java.security._
+import java.security.cert.X509Certificate
+
+import io.iohk.scalanet.codec.StreamCodec
+import io.iohk.scalanet.crypto.CryptoUtils
+import io.iohk.scalanet.crypto.CryptoUtils.Secp256r1
+import io.iohk.scalanet.monix_subject.ConnectableSubject
+import io.iohk.scalanet.peergroup.ControlEvent.InitializationError
+import io.iohk.scalanet.peergroup.InetPeerGroupUtils.toTask
+import io.iohk.scalanet.peergroup.PeerGroup.ServerEvent.ChannelCreated
+import io.iohk.scalanet.peergroup.PeerGroup.{ChannelBrokenException, HandshakeException, ServerEvent, TerminalPeerGroup}
+import io.iohk.scalanet.peergroup.dynamictls.DynamicTLSExtension.SignedKeyExtensionNodeData
+import io.iohk.scalanet.peergroup.dynamictls.DynamicTLSPeerGroup.{
+  ClientChannelImpl,
+  Config,
+  PeerInfo,
+  ServerChannelBuilder
+}
+import io.iohk.scalanet.peergroup.dynamictls.DynamicTLSPeerGroupUtils.{SSLContextForClient, SSLContextForSever}
+import io.iohk.scalanet.peergroup.{Addressable, Channel, InetMultiAddress, PeerGroup}
+import io.netty.bootstrap.{Bootstrap, ServerBootstrap}
+import io.netty.buffer.{ByteBuf, Unpooled}
+import io.netty.channel._
+import io.netty.channel.nio.NioEventLoopGroup
+import io.netty.channel.socket.SocketChannel
+import io.netty.channel.socket.nio.{NioServerSocketChannel, NioSocketChannel}
+import io.netty.handler.logging.{LogLevel, LoggingHandler}
+import io.netty.handler.ssl.{SslContext, SslHandshakeCompletionEvent}
+import javax.net.ssl._
+import monix.eval.Task
+import monix.execution.Scheduler
+import monix.reactive.observables.ConnectableObservable
+import org.bouncycastle.crypto.AsymmetricCipherKeyPair
+import org.slf4j.LoggerFactory
+import scodec.bits.BitVector
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Promise
+import scala.util.Try
+import scala.util.control.NonFatal
+
+/**
+  * PeerGroup implementation on top of TLS.
+  * the encoded bytes provided by the callers codec are not identical to the bytes put on the wire (since a
+  * length field is prepended to the byte stream). This class therefore cannot be used to talk to general services
+  * that are not instances of TLSPeerGroup.
+  *
+  * @param config bind address etc. See the companion object.
+  * @param codec  a decco codec for reading writing messages to NIO ByteBuffer.
+  * @tparam M the message type.
+  */
+class DynamicTLSPeerGroup[M](val config: Config)(
+    implicit codec: StreamCodec[M],
+    scheduler: Scheduler
+) extends TerminalPeerGroup[PeerInfo, M]() {
+
+  private val log = LoggerFactory.getLogger(getClass)
+
+  private val sslServerCtx: SslContext = DynamicTLSPeerGroupUtils.buildCustomSSlContext(SSLContextForSever, config)
+
+  private val serverSubject = ConnectableSubject[ServerEvent[PeerInfo, M]]()
+
+  private val workerGroup = new NioEventLoopGroup()
+
+  private val clientBootstrap = new Bootstrap()
```

Comment:
do we need to make it configurable?

Reply:
you mean expose most of this options to the user of library ? I think that eventually it should be done, i would not do this in this pr as it is about different thing. We should probably expose more options in other peer groups but imo it should be separate task.